llvm/lib/CodeGen/SelectionDAG/LegalizeTypes.h

```diff
@@ -796,6 +798,8 @@ class LLVM_LIBRARY_VISIBILITY DAGTypeLegalizer {
   //===--------------------------------------------------------------------===//
 
   SDValue GetSoftPromotedHalf(SDValue Op) {
+    while (Op.getNode()->getOpcode() == ISD::AssertNoFPClass)
+      Op = Op.getNode()->getOperand(0);
```

Comment:
This shouldn't be necessary

Reply:
Without it, an internal error will be trigger by `getSDValue` due to `PromotedId` is `0`.

Reply:
If this was necessary, the same would be done for AssertSext and AssertZext in their promotions, but it is not

Reply:
Sorry. Without this it will not trigger an internal error, instead some duplicated promote operation will emit like this

```
target datalayout = "e-m:e-p:32:32-Fi8-i64:64-v128:64:128-a:0:32-n32-S64"
target triple = "thumbv7em-unknown-none-eabi"

define nofpclass(nan inf) half @f1(half returned nofpclass(nan inf) %x) {
entry:
  ret half %x
}
```
will emit
```
f1:
        .fnstart
@ %bb.0:                                @ %entry
        .save   {r7, lr}
        push    {r7, lr}
        uxth    r0, r0
        bl      __aeabi_h2f
        bl      __aeabi_f2h
        pop     {r7, pc}
```

I guess there will be some same problem will be happened for `AssertSext` or `AssertZext`.